File: core/src/passes/strength_reduce.cpp
// core/src/passes/strength_reduce.cpp — strength reduction (F1-PS-011).
//
// Replaces expensive BinOps with cheaper primitives that compute the
// same value. MVP handles one pattern only:
//
//   Mul x, (1 << k)   →   Shl x, k            for k in 1..63
//
// Both directions are considered (constant left or right). The new
// Shl gets a freshly-allocated shift-count Constant; the old
// pow-of-two Constant is left for DCE to clean up.
//
// Signed/unsigned division-by-power-of-two strength reduction is
// intentionally NOT done yet — the rounding direction differs for
// signed division (we'd need a conditional rounding step) and we
// don't model signedness on BinOp size. That's future work.

#include "prisma/passes.hpp"

#include <optional>
#include <unordered_map>
#include <variant>

namespace prisma::passes {

namespace {

struct KnownConst {
    std::uint64_t value;
    ir::OpSize    size;
};

// If v is a non-zero power of two, return its log2. Otherwise nullopt.
std::optional<std::uint64_t> log2_pow2(std::uint64_t v) noexcept {
    if (v == 0 || (v & (v - 1)) != 0) return std::nullopt;
    // count trailing zeros — v is a single bit.
    std::uint64_t k = 0;
    while ((v & 1ULL) == 0) { v >>= 1; ++k; }
    return k;
}

}  // namespace

std::vector<ir::Stmt>
strength_reduce(const std::vector<ir::Stmt>& stmts) {
    std::unordered_map<ir::Ref, KnownConst> consts;

    // Find the highest ref in use so we can mint a new one for the
    // shift-count Constants this pass creates.
    ir::Ref next_ref = 0;
    for (const auto& s : stmts) {
        if (s.result && *s.result >= next_ref) next_ref = *s.result + 1;
    }

    std::vector<ir::Stmt> out;
    out.reserve(stmts.size());

    for (const auto& s : stmts) {
        // Learn Constants as we go.
        if (std::holds_alternative<ir::Constant>(s.op) && s.result) {
            const auto& c = std::get<ir::Constant>(s.op);
            consts[*s.result] = KnownConst{
                ir::mask_to_size(c.value, c.size), c.size};
        }

        if (std::holds_alternative<ir::BinOp>(s.op) && s.result) {
            const auto& b = std::get<ir::BinOp>(s.op);
            if (b.op == ir::BinOpKind::Mul) {
                // x * (1<<k)
                const auto it_r = consts.find(b.rhs);
                const auto it_l = consts.find(b.lhs);
                std::optional<std::uint64_t> k;
                ir::Ref shift_operand = 0;
                if (it_r != consts.end()) {
                    if ((k = log2_pow2(it_r->second.value))) {
                        shift_operand = b.lhs;
                    }
                } else if (it_l != consts.end()) {
                    if ((k = log2_pow2(it_l->second.value))) {
                        shift_operand = b.rhs;
                    }
                }

                if (k) {
                    // Emit a fresh Constant for the shift count, then
                    // rewrite this stmt's op to Shl.
                    const ir::Ref ref_k = next_ref++;
                    ir::Stmt k_stmt{ref_k,
                        ir::Constant{*k, b.size}};
                    out.push_back(std::move(k_stmt));
                    // Remember it so later iterations can see it.
                    consts[ref_k] = KnownConst{*k, b.size};

                    ir::Stmt rewritten{s.result,
                        ir::Op{ir::BinOp{
                            ir::BinOpKind::Shl, shift_operand, ref_k, b.size}}};
                    out.push_back(std::move(rewritten));
                    continue;
                }
            }
        }

        out.push_back(s);
    }

    return out;
}

}  // namespace prisma::passes

File: shell/prisma-passes/src/strength_reduce.rs
//! Strength reduction pass.
//!
//! Mirrors C++ `strength_reduce`. MVP pattern only:
//!
//!   `Mul x, (1 << k)`  ->  `Shl x, k`   for a power-of-two constant operand.
//!
//! A fresh `Constant` ref is minted for the shift count; the old pow-of-two
//! constant is left for DCE to clean up. Signed/unsigned division-by-pow2 is
//! intentionally not handled (rounding direction differs for signed).

use std::collections::HashMap;

use prisma_ir::{BasicBlock, BinOp, BinOpKind, Constant, Function, Op, Ref, Stmt};

use crate::Pass;

/// If `v` is a non-zero power of two, return its log2.
fn log2_pow2(v: u64) -> Option<u64> {
    if v == 0 || (v & (v - 1)) != 0 {
        return None;
    }
    Some(u64::from(v.trailing_zeros()))
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct StrengthReduce;

impl StrengthReduce {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Pass for StrengthReduce {
    fn name(&self) -> &'static str {
        "strength_reduce"
    }

    fn run(&self, func: Function) -> Function {
        strength_reduce(func)
    }
}

/// Apply strength reduction, threading a function-global ref allocator so the
/// minted shift-count constants never collide across blocks.
pub fn strength_reduce(func: Function) -> Function {
    let mut next_ref: Ref = func
        .blocks
        .iter()
        .flat_map(|b| b.stmts.iter())
        .filter_map(|s| s.result)
        .map(|r| r + 1)
        .max()
        .unwrap_or(0);

    let blocks = func
        .blocks
        .into_iter()
        .map(|block| {
            let mut consts: HashMap<Ref, u64> = HashMap::new();
            let mut out = Vec::with_capacity(block.stmts.len());

            for stmt in block.stmts {
                if let (Some(result), Op::Constant(c)) = (stmt.result, &stmt.op) {
                    consts.insert(result, c.value & c.size.mask());
                }

                if let (Some(result), Op::BinOp(b)) = (stmt.result, &stmt.op) {
                    if b.op == BinOpKind::Mul {
                        let (k, shift_operand) = match (
                            consts.get(&b.rhs).and_then(|&v| log2_pow2(v)),
                            consts.get(&b.lhs).and_then(|&v| log2_pow2(v)),
                        ) {
                            (Some(k), _) => (Some(k), b.lhs),
                            (None, Some(k)) => (Some(k), b.rhs),
                            (None, None) => (None, 0),
                        };

                        if let Some(k) = k {
                            let ref_k = next_ref;
                            next_ref += 1;
                            consts.insert(ref_k, k);
                            out.push(Stmt::new(
                                Some(ref_k),
                                Op::Constant(Constant { value: k, size: b.size }),
                            ));
                            out.push(Stmt::new(
                                Some(result),
                                Op::BinOp(BinOp {
                                    op: BinOpKind::Shl,
                                    lhs: shift_operand,
                                    rhs: ref_k,
                                    size: b.size,
                                }),
                            ));
                            continue;
                        }
                    }
                }

                out.push(stmt);
            }

            BasicBlock {
                id: block.id,
                stmts: out,
            }
        })
        .collect();

    Function {
        blocks,
        entry: func.entry,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use prisma_ir::OpSize;

    #[test]
    fn mul_by_pow2_becomes_shift() {
        // r0 = 8 ; r1 = mul r9, r0   ->   r0=8 ; r2=3 ; r1 = shl r9, r2
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(Some(0), Op::Constant(Constant { value: 8, size: OpSize::I64 })),
                    Stmt::new(
                        Some(1),
                        Op::BinOp(BinOp { op: BinOpKind::Mul, lhs: 9, rhs: 0, size: OpSize::I64 }),
                    ),
                ],
            }],
        };
        let out = strength_reduce(func);
        let stmts = &out.blocks[0].stmts;
        assert_eq!(stmts.len(), 3);
        // minted shift count = log2(8) = 3
        match &stmts[1].op {
            Op::Constant(c) => assert_eq!(c.value, 3),
            other => panic!("expected minted constant, got {other:?}"),
        }
        match &stmts[2].op {
            Op::BinOp(b) => {
                assert_eq!(b.op, BinOpKind::Shl);
                assert_eq!(b.lhs, 9);
                assert_eq!(b.rhs, stmts[1].result.unwrap());
            }
            other => panic!("expected shl, got {other:?}"),
        }
    }

    #[test]
    fn mul_by_non_pow2_is_untouched() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(Some(0), Op::Constant(Constant { value: 6, size: OpSize::I64 })),
                    Stmt::new(
                        Some(1),
                        Op::BinOp(BinOp { op: BinOpKind::Mul, lhs: 9, rhs: 0, size: OpSize::I64 }),
                    ),
                ],
            }],
        };
        let out = strength_reduce(func.clone());
        assert_eq!(out, func);
    }

    #[test]
    fn idempotent() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(Some(0), Op::Constant(Constant { value: 16, size: OpSize::I32 })),
                    Stmt::new(
                        Some(1),
                        Op::BinOp(BinOp { op: BinOpKind::Mul, lhs: 0, rhs: 9, size: OpSize::I32 }),
                    ),
                ],
            }],
        };
        let once = strength_reduce(func);
        let twice = strength_reduce(once.clone());
        assert_eq!(once, twice);
    }
}
